substrate/client/network/sync/src/lib.rs

```diff
@@ -407,6 +407,122 @@ impl<B: BlockT> PeerSyncState<B> {
+	downloaded: Vec<BlockData<B>>,
+}
+
+impl<B: BlockT> PeerDownloadState<B> {
+	fn new(common_number: NumberFor<B>, range: Range<NumberFor<B>>) -> Self {
+		Self { common_number, range, downloaded: Vec::new() }
+	}
+
+	/// Handles the new blocks received from the peer.
+	/// Returns:
+	/// Ok(true): If done and no more requests need to be issued.
+	/// Ok(false): if more requests need to be issued.
+	/// Err(): On any failures.
+	fn handle_blocks(
+		&mut self,
+		who: &PeerId,
+		mut new_blocks: Vec<BlockData<B>>,
+		request: BlockRequest<B>,
+		collection: &mut BlockCollection<B>,
+	) -> Result<bool, BadPeer> {
+		collection.clear_peer_download(who);
+
+		// Validate the blocks.
+		let start_block = match validate_blocks::<B>(&new_blocks, who, Some(request)) {
+			Ok(Some(start_block)) => start_block,
+			Ok(None) => return Ok(true),
+			Err(err) => return Err(err),
+		};
+
+		// Add the new blocks to the downloaded list.
+		let range_len = (self.range.end - self.range.start).saturated_into::<u32>() as usize;
+		if (self.downloaded.len() + new_blocks.len()) > range_len {
+			return Err(BadPeer(*who, rep::BAD_RESPONSE))
+		}
+		let new_blocks_len = new_blocks.len();
+		new_blocks.append(&mut self.downloaded);
+		self.downloaded.append(&mut new_blocks);
+
+		if self.range.start == (self.common_number + One::one()) &&
+			self.downloaded.len() < range_len
+		{
+			// The download extends the common ancestor, but the response
+			// was partial. We want to issue further requests to fetch
+			// the incomplete part before importing.
+			trace!(
+				target: LOG_TARGET,
+				"Download state: incomplete download: \
+				common = {}, range = {:?}, new_blocks = {new_blocks_len}, downloaded = {}",
+				self.common_number, self.range, self.downloaded.len()
+			);
+			return Ok(false)
+		}
+
+		// Done, report the accumulated blocks to the collection.
+		let mut downloaded = Vec::new();
+		downloaded.append(&mut self.downloaded);
+		collection.insert(start_block, downloaded, *who);
+		Ok(true)
+	}
+
+	/// Returns the next request to be issued based on the download state.
+	fn peer_block_request(
+		&self,
+		attrs: BlockAttributes,
+		peer_best_number: NumberFor<B>,
+		peer_best_hash: &B::Hash,
+	) -> Option<(Range<NumberFor<B>>, BlockRequest<B>)> {
+		let downloaded_head = if let Some(number) =
+			self.downloaded.first().and_then(|b| b.header.as_ref()).map(|h| *h.number())
+		{
+			number
+		} else {
+			warn!(
+				target: LOG_TARGET,
+				"Download state: unable to get block number of the downloaded blocks: \
+				common = {}, range = {:?}, downloaded = {}",
+				self.common_number, self.range, self.downloaded.len()
+			);
+			return None
+		};
+		let range = Range { start: self.range.start, end: downloaded_head };
+
+		// The end is not part of the range.
+		let last = range.end.saturating_sub(One::one());
+		let from = if peer_best_number == last {
+			FromBlock::Hash(*peer_best_hash)
+		} else {
+			FromBlock::Number(last)
+		};
```

Comment:
This is existing behavior, just retained as is: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/network/sync/src/lib.rs#L2561

Reply:
The `request.from` is the hash of the first block we want to download, which we don't know until the block is downloaded (now I understand why current code sets it hash if `block_number == peer.best_block_number`, in this case we happen to know the corresponding `peer.best_block_hash`)

Unless we want to set it to parent hash of the first block already downloaded. I would prefer to keep this uniform with existing code, don't know why current code does not set it to parent hash.

Reply:
Yes, I meant setting it to the parent hash of the already downloaded first block in the range. Otherwise we can't be sure that we are downloading the right fork and can fail when importing the range.

Reply:
Changed it to pass parent hash